packages/snfoundry/scripts-ts/helpers/fees.ts

```diff
@@ -0,0 +1,94 @@
+import { Contract, Provider, uint256 } from "starknet";
+import { Abi } from "starknet";
+import { red, yellow } from "./colorize-log";
+import { Network } from "../types";
+
+export const erc20ABI = [
+    {
+        inputs: [
+            {
+                name: "account",
+                type: "felt",
+            },
+        ],
+        name: "balanceOf",
+        outputs: [
+            {
+                name: "balance",
+                type: "Uint256",
+            },
+        ],
+        stateMutability: "view",
+        type: "function",
+    },
+] satisfies Abi;
+
+//function to decide preferred token for fee payment
+export async function getTxVersion(
+    network: Network,
+    feeToken: string,
+) {
+    const { feeToken: feeTokenOptions, provider, deployer } = network;
+
+    //check the specified feeToken
+    const specifiedToken = feeTokenOptions.find(
+        (token) => token.name === feeToken
+    );
+    if (specifiedToken) {
+        const balance = await getBalance(
+            deployer.address,
+            provider,
+            specifiedToken.address
+        );
+        if (balance > 0n) {
+            console.log(yellow(`Using ${feeToken.toUpperCase()} as fee token`));
+            return getTxVersionFromFeeToken(feeToken);
+        }
+        console.log(
+            red(`${feeToken.toUpperCase()} balance is zero, trying other options`)
+        );
+    }
+
+    // Check other options
+    for (const token of feeTokenOptions) {
+        if (token.name !== feeToken) {
+            const balance = await getBalance(
+                deployer.address,
+                provider,
+                token.address
+            );
+            if (balance > 0n) {
+                console.log(yellow(`Using ${token.name.toUpperCase()} as fee token`));
+                return getTxVersionFromFeeToken(token.name);
+            }
+            console.log(
+                red(`${token.name.toUpperCase()} balance is zero, trying next option`)
+            );
+        }
+    }
+
+    console.error(
+        red("Error: Unable to find a fee token with sufficient balance. Please fund your wallet first.")
+    );
+    throw new Error("No fee token with balance found");
+}
+
+export async function getBalance(
+    account: string,
+    provider: Provider,
+    tokenAddress: string
+): Promise<bigint> {
+    try {
+        const contract = new Contract(erc20ABI, tokenAddress, provider);
+        const { balance } = await contract.balanceOf(account);
+        return uint256.uint256ToBN(balance);
+    } catch (error) {
+        console.error('Error fetching balance:', error);
+        throw error;
+    }
+   
+}
+
+function getTxVersionFromFeeToken(feeToken: string) {
+    return feeToken === "strk" ? "0x3" : "0x1";
```

Comment:
@metalboyrick @Jemiiah this doesnt work on my side following the right versions on README

i get the following error when doing:

`yarn chain` , `yarn deploy`

```
-32602: Invalid declare transaction v1: missing field `offset`: undefined
```

The declare transaction fails on a fresh clone. This can be caused due to a misunderstanding on transaction version on contract declaration, please take a look [here](https://github.com/starknet-io/starknet.js/blob/b15fe2d6aa9a5494c510fcac3d6d34fd7e41bf5c/src/account/default.ts#L393C2-L416C7) : 

- v1 : not using sierra ( old )
- v2 : legacy uses sierra and paid in ETH
- v3 : legacy uses sierra and paid in STRK

We dont support tx version `0x1` on the framework as we always need a sierra compiled file [here](https://github.com/Scaffold-Stark/scaffold-stark-2/blob/3b9562df48613e5cb37bcbdb427ca31c57d350b0/packages/snfoundry/scripts-ts/deploy-contract.ts#L169C3-L186C4)

Reply:
the `--fee` parameter works ok